From c3eaa6fee0a20737c3988f4af47f1f449e456f7a Mon Sep 17 00:00:00 2001
From: Ricardo Garcia <rgarcia@igalia.com>
Date: Thu, 14 May 2026 11:16:44 +0200
Subject: [PATCH] Fix validation errors in copy memory indirect tests

* Missing VK_BUFFER_USAGE_INDIRECT_BUFFER_BIT in several buffers.
* Missing VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT in several buffers.
* Make sure dedicated allocations include the proper allocation flags.
  This is needed when combining dedicated allocations with device
  addresses. Unifies logic in vkMemUtil around these flags.

Affects:
dEQP-VK.api.copy_and_blit.*.memory_to_image_indirect*

Potentially other tests could be affected due to the changes in
vkMemUtil, but the tests above already exercise the proper paths.

Components: Vulkan
VK-GL-CTS issue: 6423

Upstream-Status: Backport [https://github.com/KhronosGroup/VK-GL-CTS/commit/c3eaa6fee0a20737c3988f4af47f1f449e456f7a]
Change-Id: Ia44702489a9c5551a00900e328d2c19e5cbb0853
---
 .../framework/vulkan/vkImageUtil.cpp          | 12 +--
 .../vulkancts/framework/vulkan/vkMemUtil.cpp  | 79 ++++++++++++-------
 .../api/vktApiCopiesAndBlittingTests.cpp      | 17 ++--
 3 files changed, 71 insertions(+), 37 deletions(-)

diff --git a/external/vulkancts/framework/vulkan/vkImageUtil.cpp b/external/vulkancts/framework/vulkan/vkImageUtil.cpp
index 531f87246..c8cd94916 100644
--- a/external/vulkancts/framework/vulkan/vkImageUtil.cpp
+++ b/external/vulkancts/framework/vulkan/vkImageUtil.cpp
@@ -4756,17 +4756,19 @@ void copyBufferToImageIndirect(const DeviceInterface &vk, const InstanceInterfac
 
     // Create indirect buffer for copy commands
     const VkDeviceSize indirectBufferSize = copyRegions.size() * sizeof(VkCopyMemoryToImageIndirectCommandKHR);
+    const auto indirectBufferUsage =
+        static_cast<VkBufferUsageFlags>(VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT |
+                                        VK_BUFFER_USAGE_INDIRECT_BUFFER_BIT);
 
     const VkBufferCreateInfo indirectBufferCreateInfo = {
         VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO, // VkStructureType sType
         nullptr,                              // const void* pNext
         0u,                                   // VkBufferCreateFlags flags
         indirectBufferSize,                   // VkDeviceSize size
-        VK_BUFFER_USAGE_TRANSFER_SRC_BIT |    // VkBufferUsageFlags usage
-            VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT,
-        VK_SHARING_MODE_EXCLUSIVE, // VkSharingMode sharingMode
-        0u,                        // uint32_t queueFamilyIndexCount
-        nullptr                    // const uint32_t* pQueueFamilyIndices
+        indirectBufferUsage,                  // VkBufferUsageFlags usage
+        VK_SHARING_MODE_EXCLUSIVE,            // VkSharingMode sharingMode
+        0u,                                   // uint32_t queueFamilyIndexCount
+        nullptr                               // const uint32_t* pQueueFamilyIndices
     };
 
     Move<VkBuffer> indirectBuffer = createBuffer(vk, device, &indirectBufferCreateInfo);
diff --git a/external/vulkancts/framework/vulkan/vkMemUtil.cpp b/external/vulkancts/framework/vulkan/vkMemUtil.cpp
index b7e35e3d0..6bd036b17 100644
--- a/external/vulkancts/framework/vulkan/vkMemUtil.cpp
+++ b/external/vulkancts/framework/vulkan/vkMemUtil.cpp
@@ -234,6 +234,41 @@ MovePtr<Allocation> SimpleAllocator::allocate(const VkMemoryAllocateInfo &allocI
     return MovePtr<Allocation>(new SimpleAllocation(mem, hostPtr, static_cast<size_t>(offset)));
 }
 
+namespace
+{
+
+std::unique_ptr<VkMemoryAllocateFlagsInfo> getMemoryAllocateFlagsInfo(MemoryRequirement requirement)
+{
+    VkMemoryAllocateFlags flags = 0u;
+
+    if (requirement & MemoryRequirement::DeviceAddress)
+        flags |= VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT;
+
+    if (requirement & MemoryRequirement::DeviceAddressCaptureReplay)
+        flags |= VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_CAPTURE_REPLAY_BIT;
+
+#ifndef CTS_USES_VULKANSC
+    if (requirement & MemoryRequirement::ZeroInitialize)
+        flags |= VK_MEMORY_ALLOCATE_ZERO_INITIALIZE_BIT_EXT;
+#endif // CTS_USES_VULKANSC
+
+    std::unique_ptr<VkMemoryAllocateFlagsInfo> flagsInfo;
+
+    if (flags)
+    {
+        flagsInfo.reset(new VkMemoryAllocateFlagsInfo{
+            VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_FLAGS_INFO, // VkStructureType       sType
+            nullptr,                                      // const void*           pNext
+            flags,                                        // VkMemoryAllocateFlags flags
+            0u,                                           // uint32_t              deviceMask
+        });
+    }
+
+    return flagsInfo;
+}
+
+} // namespace
+
 MovePtr<Allocation> SimpleAllocator::allocate(const VkMemoryRequirements &memReqs, MemoryRequirement requirement,
                                               const tcu::Maybe<HostIntent> &hostIntent,
                                               uint64_t memoryOpaqueCaptureAddr)
@@ -258,12 +293,7 @@ MovePtr<Allocation> SimpleAllocator::allocate(const VkMemoryRequirements &memReq
         memoryTypeNdx,                          // uint32_t memoryTypeIndex;
     };
 
-    VkMemoryAllocateFlagsInfo allocFlagsInfo = {
-        VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_FLAGS_INFO, //    VkStructureType            sType
-        nullptr,                                      //    const void*                pNext
-        0,                                            //    VkMemoryAllocateFlags    flags
-        0,                                            //    uint32_t                deviceMask
-    };
+    auto allocFlagsInfo = getMemoryAllocateFlagsInfo(requirement);
 
     VkMemoryOpaqueCaptureAddressAllocateInfo captureInfo = {
         VK_STRUCTURE_TYPE_MEMORY_OPAQUE_CAPTURE_ADDRESS_ALLOCATE_INFO, // VkStructureType sType
@@ -271,24 +301,11 @@ MovePtr<Allocation> SimpleAllocator::allocate(const VkMemoryRequirements &memReq
         memoryOpaqueCaptureAddr,                                       // uint64_t        opaqueCaptureAddress
     };
 
-    if (requirement & MemoryRequirement::DeviceAddress)
-        allocFlagsInfo.flags |= VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT;
+    if (memoryOpaqueCaptureAddr && allocFlagsInfo)
+        allocFlagsInfo->pNext = &captureInfo;
 
-    if (requirement & MemoryRequirement::DeviceAddressCaptureReplay)
-    {
-        allocFlagsInfo.flags |= VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_CAPTURE_REPLAY_BIT;
-
-        if (memoryOpaqueCaptureAddr)
-            allocFlagsInfo.pNext = &captureInfo;
-    }
-
-#ifndef CTS_USES_VULKANSC
-    if (requirement & MemoryRequirement::ZeroInitialize)
-        allocFlagsInfo.flags |= VK_MEMORY_ALLOCATE_ZERO_INITIALIZE_BIT_EXT;
-#endif // CTS_USES_VULKANSC
-
-    if (allocFlagsInfo.flags)
-        allocInfo.pNext = &allocFlagsInfo;
+    if (allocFlagsInfo)
+        allocInfo.pNext = allocFlagsInfo.get();
 
     Move<VkDeviceMemory> mem = allocateMemory(m_vk, m_device, &allocInfo);
     MovePtr<HostPtr> hostPtr;
@@ -361,14 +378,18 @@ de::MovePtr<Allocation> allocateDedicated(const InstanceInterface &vki, const De
                                           const VkPhysicalDevice &physDevice, const VkDevice device,
                                           const VkBuffer buffer, MemoryRequirement requirement)
 {
-    const VkMemoryRequirements memoryRequirements               = getBufferMemoryRequirements(vkd, device, buffer);
-    const VkMemoryDedicatedAllocateInfo dedicatedAllocationInfo = {
+    const VkMemoryRequirements memoryRequirements         = getBufferMemoryRequirements(vkd, device, buffer);
+    VkMemoryDedicatedAllocateInfo dedicatedAllocationInfo = {
         VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO, // VkStructureType        sType
         nullptr,                                          // const void*            pNext
         VK_NULL_HANDLE,                                   // VkImage                image
         buffer                                            // VkBuffer                buffer
     };
 
+    const auto flagsInfo = getMemoryAllocateFlagsInfo(requirement);
+    if (flagsInfo)
+        dedicatedAllocationInfo.pNext = flagsInfo.get();
+
     return allocateExtended(vki, vkd, physDevice, device, memoryRequirements, requirement, &dedicatedAllocationInfo);
 }
 
@@ -376,14 +397,18 @@ de::MovePtr<Allocation> allocateDedicated(const InstanceInterface &vki, const De
                                           const VkPhysicalDevice &physDevice, const VkDevice device,
                                           const VkImage image, MemoryRequirement requirement)
 {
-    const VkMemoryRequirements memoryRequirements               = getImageMemoryRequirements(vkd, device, image);
-    const VkMemoryDedicatedAllocateInfo dedicatedAllocationInfo = {
+    const VkMemoryRequirements memoryRequirements         = getImageMemoryRequirements(vkd, device, image);
+    VkMemoryDedicatedAllocateInfo dedicatedAllocationInfo = {
         VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO, // VkStructureType        sType
         nullptr,                                          // const void*            pNext
         image,                                            // VkImage                image
         VK_NULL_HANDLE                                    // VkBuffer                buffer
     };
 
+    const auto flagsInfo = getMemoryAllocateFlagsInfo(requirement);
+    if (flagsInfo)
+        dedicatedAllocationInfo.pNext = flagsInfo.get();
+
     return allocateExtended(vki, vkd, physDevice, device, memoryRequirements, requirement, &dedicatedAllocationInfo);
 }
 
diff --git a/external/vulkancts/modules/vulkan/api/vktApiCopiesAndBlittingTests.cpp b/external/vulkancts/modules/vulkan/api/vktApiCopiesAndBlittingTests.cpp
index 5bd0e819e..5ace0c58f 100644
--- a/external/vulkancts/modules/vulkan/api/vktApiCopiesAndBlittingTests.cpp
+++ b/external/vulkancts/modules/vulkan/api/vktApiCopiesAndBlittingTests.cpp
@@ -16873,19 +16873,24 @@ CopyMemoryToImageIndirect::CopyMemoryToImageIndirect(Context &context, TestParam
 
     // Create source buffer
     {
+        const auto sourceBufferUsage = static_cast<VkBufferUsageFlags>(VK_BUFFER_USAGE_TRANSFER_SRC_BIT |
+                                                                       VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT);
+
         const VkBufferCreateInfo sourceBufferParams = {
             VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO, // VkStructureType sType;
             nullptr,                              // const void* pNext;
             0u,                                   // VkBufferCreateFlags flags;
             m_bufferSize,                         // VkDeviceSize size;
-            VK_BUFFER_USAGE_TRANSFER_SRC_BIT,     // VkBufferUsageFlags usage;
+            sourceBufferUsage,                    // VkBufferUsageFlags usage;
             VK_SHARING_MODE_EXCLUSIVE,            // VkSharingMode sharingMode;
             0u,                                   // uint32_t queueFamilyIndexCount;
             (const uint32_t *)nullptr,            // const uint32_t* pQueueFamilyIndices;
         };
 
+        const auto sourceBufferMemReqs = (MemoryRequirement::DeviceAddress | MemoryRequirement::HostVisible);
+
         m_source            = createBuffer(vk, m_device, &sourceBufferParams);
-        m_sourceBufferAlloc = allocateBuffer(vki, vk, vkPhysDevice, m_device, *m_source, MemoryRequirement::HostVisible,
+        m_sourceBufferAlloc = allocateBuffer(vki, vk, vkPhysDevice, m_device, *m_source, sourceBufferMemReqs,
                                              *m_allocator, m_params.allocationKind);
         VK_CHECK(vk.bindBufferMemory(m_device, *m_source, m_sourceBufferAlloc->getMemory(),
                                      m_sourceBufferAlloc->getOffset()));
@@ -17151,9 +17156,11 @@ tcu::TestStatus CopyMemoryToImageIndirect::iterate(void)
     };
 
     const VkDeviceSize indirectBufferSize = de::max(m_params.regions.size(), (size_t)1) * sizeof(IndirectImageParams);
-    const BufferWithMemory indirectBuffer(
-        vk, vkDevice, memAlloc, makeBufferCreateInfo(indirectBufferSize, VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT),
-        MemoryRequirement::HostVisible | MemoryRequirement::DeviceAddress);
+    const auto indirectBufferUsage        = static_cast<VkBufferUsageFlags>(VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT |
+                                                                     VK_BUFFER_USAGE_INDIRECT_BUFFER_BIT);
+    const BufferWithMemory indirectBuffer(vk, vkDevice, memAlloc,
+                                          makeBufferCreateInfo(indirectBufferSize, indirectBufferUsage),
+                                          MemoryRequirement::HostVisible | MemoryRequirement::DeviceAddress);
 
     // indirectBuffer Address
     VkBufferDeviceAddressInfo indirectBufferAddressInfo{VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO, nullptr,
-- 
2.43.0

